### PR TITLE
osd: make snapmapper warn+clean up instead of assert

### PR DESCRIPTION
**Pull request by liewegas**
http://tracker.ceph.com/issues/22752

### CODE REVIEW

File: src/osd/PG.cc [hunk 1]
```diff
@@ -4182,9 +4182,10 @@ void PG::_scan_snaps(ScrubMap &smap)
 			    << "...repaired";
 	}
 	snap_mapper.add_oid(hoid, obj_snaps, &_t);
-	r = osd->store->apply_transaction(osr.get(), std::move(t));
+	// we don't particularly care when this is visible or committed.
+	r = osd->store->queue_transaction(osr.get(), std::move(t), nullptr);
```
Comment (gregsfortytwo) [inline src/osd/PG.cc]: Do we need to worry about skipping a snaptrim, because it gets processed before the repair is complete?

**Comment by gregsfortytwo** [comment]
This looks good to me, but it's still marked WIP so I'm not marking it as approved. Feel free to do so when ready!

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-14_04:42:43-rados-wip-kefu-testing-2018-03-14-1106-distro-basic-mira/

- SLOW_OPS failures are introduced by #20660
- the coredump is addressed by #20909

File: src/osd/SnapMapper.cc [hunk 2]
```diff
@@ -243,6 +243,7 @@ void SnapMapper::add_oid(
   MapCacher::Transaction<std::string, bufferlist> *t)
 {
   dout(20) << __func__ << " " << oid << " " << snaps << dendl;
+  assert(!snaps.empty());
```
Comment (tchaikov) [inline src/osd/SnapMapper.cc]: if this is the case, we might want to `assert(0 == "decode snaps failure")` at https://github.com/ceph/ceph/blob/master/src/osd/PG.cc#L3747 in `PG::update_snap_map()`.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-28_14:42:54-rados-wip-kefu-testing-2018-03-28-1948-distro-basic-smithi/

**Pull request merged**